artiq/gateware/drtio/transceiver/eem_serdes.py

```diff
@@ -0,0 +1,408 @@
+                    o_CNTVALUEOUT=self.cnt_out[i],
+
+                    i_IDATAIN=self.ser_in_no_dly[i],
+                    o_DATAOUT=ser_in[i]
+                ),
+
+                # IDELAYCTRL is with the clocking
+            ]
+
+
+class TXSerdes(Module):
+    def __init__(self):
+        self.txdata = [ Signal(5) for _ in range(4) ]
+        self.ser_out = [ Signal() for _ in range(4) ]
+        self.t_out = [ Signal() for _ in range(4) ]
+
+        # TX SERDES
+        for i in range(4):
+            self.specials += Instance("OSERDESE2",
+                p_DATA_RATE_OQ="SDR", p_DATA_RATE_TQ="BUF",
+                p_DATA_WIDTH=5, p_TRISTATE_WIDTH=1,
+                p_INIT_OQ=0b00000,
+                o_OQ=self.ser_out[i],
+                o_TQ=self.t_out[i],
+                i_RST=ResetSignal(),
+                i_CLK=ClockSignal("sys5x"),
+                i_CLKDIV=ClockSignal(),
+                i_D1=self.txdata[i][0],
+                i_D2=self.txdata[i][1],
+                i_D3=self.txdata[i][2],
+                i_D4=self.txdata[i][3],
+                i_D5=self.txdata[i][4],
+                i_TCE=1, i_OCE=1,
+                i_T1=0)
+
+
+class MultiEncoder(Module):
+    def __init__(self):
+        WORDS = 2
+        # Keep the link layer interface identical to standard encoders
+        self.d = [Signal(8) for _ in range(WORDS)]
+        self.k = [Signal() for _ in range(WORDS)]
+
+        # Output interface is simplified because we have custom physical layer
+        self.output = [Signal(10) for _ in range(WORDS)]
+
+        # Phase of the encoder
+        # Alternate crossbar between encoder and SERDES every cycle
+        self.phase = Signal()
+
+        # Intermediate registers for output and disparity
+        # More significant bits are buffered due to channel geometry
+        # Disparity bit is delayed. The same encoder is shared by 2 SERDES
+        output_bufs = [Signal(5) for _ in range(WORDS)]
+        disp_bufs = [Signal() for _ in range(WORDS)]
+
+        encoders = [SingleEncoder() for _ in range(WORDS)]
+        self.submodules += encoders
+
+        for d, k, output, output_buf, disp_buf, encoder in \
+                zip(self.d, self.k, self.output, output_bufs, disp_bufs, encoders):
+            self.comb += [
+                encoder.d.eq(d),
+                encoder.k.eq(k),
+
+                # Implementing switching crossbar
+                If(self.phase,
+                    output.eq(Cat(encoder.output[0:5], output_buf))
+                ).Else(
+                    output.eq(Cat(output_buf, encoder.output[0:5]))
+                ),
+            ]
+            # Handle intermediate registers
+            self.sync += [
+                disp_buf.eq(encoder.disp_out),
+                encoder.disp_in.eq(disp_buf),
+                output_buf.eq(encoder.output[5:10]),
+            ]
+
+
+# Unlike the usual 8b10b decoder, it needs to know which SERDES to decode
+class CrossbarDecoder(Module):
```

Comment:
I thought of such because it switches sources while maintaining an output. It is really just alternating between 2 states though. I have changed it to `MultiDecoder` to match the encoder naming.